cpp/src/bitmask/bit_mask.cuh

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _BIT_MASK_H_
+#define _BIT_MASK_H_
+
+#include "cudf.h"
+#include "bit_mask.h"
+#include "rmm/rmm.h"
+#include "utilities/error_utils.h"
+#include <cuda_runtime_api.h>
+
+using bit_mask_t = bit_mask::bit_mask_t;
+
+
+/* ---------------------------------------------------------------------------- */
+/**
+ * @brief  Class for managing bit containers on the device
+ */
+/* ---------------------------------------------------------------------------- */
+class BitMask {
```

Comment:
Should `BitMask` have `operator[]()` that calls `get_element` or `set_element` as appropriate?

Reply:
I could add the operator as a device-only function, which should just return a reference to the appropriate element.

Adding it as a host function (using get_element/set_element) would, I believe, require wrapping the element in another class, or relying on the caller understanding that they need to call get_element/set_element on the returned pointer.

Reply:
Thought about it a bit more.  The GetBit and SetBit methods operate on a single bit.  The GetElement and SetElement methods operate on a single element.  Would overloading the [] operator and having it operate at the element level be intuitive?

I think I'd be more comfortable adding a GetElement and SetElement to the class that take an element index rather than an ambiguous operator overload.

Reply:
Good point. I think you are right we shouldn't have `operator[]` due to possible confusion.